### PR TITLE
[ADD] Support `BCEWithLogitsLoss` in Fisher-MC and KFAC

### LINKED ISSUE #90: Support `BCEWithLogitsLoss`
Requested by @wiseodd for `laplace-torch`.

This consists of three parts:

- [x] Support `sqrt_hessian` for `KFAC` (type 2)
- [x] Support `sample_grad_output` for `FisherMC`
- [x] Support `draw_label` for `KFAC`  (type 1)

### PR DESCRIPTION
**Pull request by f-dangel**
This makes `curvlinops` support binary classification. Resolves #90

### CODE REVIEW

File: curvlinops/fisher.py [hunk 1]
```diff
@@ -253,6 +244,7 @@ def sample_grad_output(self, output: Tensor, num_samples: int) -> Tensor:
             output: model prediction ``f`` for multiple data with batch axis as
                 0th dimension.
             num_samples: Number of samples to draw.
+            y: Ground truth.
```
Comment (runame) [inline curvlinops/fisher.py:247]: I think I would prefer another term here, something like "empirical label" or "label from data". "Ground truth" implies to me that this is what we are trying to approximate or what would be the ideal object to use.

File: curvlinops/kfac.py [hunk 2]
```diff
@@ -555,6 +563,18 @@ def _compute_loss_and_backward(self, output: Tensor, y: Tensor):
...
+
+                if (
+                    isinstance(self._loss_func, (BCEWithLogitsLoss, MSELoss))
+                    and self._loss_func.reduction == "mean"
+                ):
+                    # ``BCEWithLogitsLoss`` and ``MSELoss`` also average over non-batch
+                    # dimensions. We have to scale the loss to incorporate this scaling
+                    # as we cannot generally achieve it by incorporating it into the
+                    # drawn sample.
+                    C = output.shape[1:].numel()
```
Comment (runame) [inline curvlinops/kfac.py:575]: Should `output.shape[1]` be sufficient here since we always reshape the outputs to be 2d?

**Review by runame — APPROVED** [review]
Looks good to me!

File: curvlinops/fisher.py (hunk 1 above)
Comment (f-dangel) [reply curvlinops/fisher.py:247]: Agree.

File: curvlinops/kfac.py (hunk 2 above)
Comment (f-dangel) [reply curvlinops/kfac.py:575]: Made more explicit.

**Pull request merged**